apps/shared/avifpng.c

```diff
@@ -119,46 +95,107 @@ static avifBool avifCopyRawProfile(const char * profile, size_t profileLength, a
+// Returns AVIF_FALSE in case of a parsing error.
+static avifBool avifExtractExifAndXMP(png_structp png, png_infop info, avifBool * ignoreExif, avifBool * ignoreXMP, avifImage * avif)
 {
 #ifdef PNG_eXIf_SUPPORTED
-    png_uint_32 exifSize = 0;
-    png_bytep exif = NULL;
-    if (png_get_eXIf_1(png, info, &exifSize, &exif) == PNG_INFO_eXIf) {
-        if ((exifSize == 0) || !exif) {
-            fprintf(stderr, "Exif extraction failed: empty eXIf chunk\n");
-            return AVIF_FALSE;
+    if (!*ignoreExif) {
+        png_uint_32 exifSize = 0;
+        png_bytep exif = NULL;
+        if (png_get_eXIf_1(png, info, &exifSize, &exif) == PNG_INFO_eXIf) {
+            if ((exifSize == 0) || !exif) {
+                fprintf(stderr, "Exif extraction failed: empty eXIf chunk\n");
+                return AVIF_FALSE;
+            }
+            avifImageSetMetadataExif(avif, exif, exifSize);
+            *ignoreExif = AVIF_TRUE; // Ignore any other Exif chunk.
         }
-        avifImageSetMetadataExif(avif, exif, exifSize);
-        return AVIF_TRUE;
     }
 #endif // PNG_eXIf_SUPPORTED
 
+    // tXMP could be retrieved using the png_get_unknown_chunks() API but tXMP is deprecated
+    // and there is no PNG file example with a tXMP chunk lying around, so it is not worth the hassle.
+
     png_textp text = NULL;
     const png_uint_32 numTextChunks = png_get_text(png, info, &text, NULL);
-    for (png_uint_32 i = 0; i < numTextChunks; ++i, ++text) {
-        if (!strcmp(text->key, "Raw profile type exif") || !strcmp(text->key, "Raw profile type APP1")) {
-            png_size_t textLength;
-            switch (text->compression) {
+    for (png_uint_32 i = 0; (!*ignoreExif || !*ignoreXMP) && (i < numTextChunks); ++i, ++text) {
+        png_size_t textLength = text->text_length;
 #ifdef PNG_iTXt_SUPPORTED
-                case PNG_ITXT_COMPRESSION_NONE:
-                case PNG_ITXT_COMPRESSION_zTXt:
-                    textLength = text->itxt_length;
-                    break;
+        if ((text->compression == PNG_ITXT_COMPRESSION_NONE) || (text->compression == PNG_ITXT_COMPRESSION_zTXt)) {
+            textLength = text->itxt_length;
+        }
 #endif
-                case PNG_TEXT_COMPRESSION_NONE:
-                case PNG_TEXT_COMPRESSION_zTXt:
-                default:
-                    textLength = text->text_length;
-                    break;
+
+        if (!*ignoreExif && !strcmp(text->key, "Raw profile type exif")) {
+            if (!avifCopyRawProfile(text->text, textLength, &avif->exif)) {
+                return AVIF_FALSE;
+            }
+            avifRemoveHeadingExif00TagIfAny(&avif->exif);
+            *ignoreExif = AVIF_TRUE; // Ignore any other Exif chunk.
+        } else if (!*ignoreXMP && !strcmp(text->key, "Raw profile type xmp")) {
+            if (!avifCopyRawProfile(text->text, textLength, &avif->xmp)) {
```

Comment:
Don't see where the potential `http://ns.adobe.com/xap/1.0/\0` header is handled in this case?